### PR TITLE
[4.0] Mail Templates improvement

### PR DESCRIPTION
**Pull request by joomdonation**
Pull Request for Issue #27619.

### Summary of Changes
This PR makes some improvements to Mail Templates to make it easier to use (I hope :D ). Below are the changes (base on suggestions from https://github.com/joomla/joomla-cms/issues/27619)

1. Link Title column to edit mail template of the default language
2. Remove **Templates** and No **Templates** columns
3. On a site with multiple content languages, add **Edit Templates** column to allow click on to edit mail template of a specific language
4. Remove switcher, show actual mail template content instead of just language item
5. Add **Reset To Default** to allow reset to default subject, body and html body of a mail template
6. Replace **Available Tags** with **Insert Tags**
7. Show Default Content For HTML Body (base on body with nl2br method applied) if there is no data for HTML Body from master template

Below are some screenshot to show the changes:

- Mail Templates on a single/mono language
mono-language

- Mail Templates on multiple languages
multiple_languages

- Edit Mail Template of a language
edit_mail_template

### Testing Instructions
Download update package for this PR at https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/33230/downloads/42268/Joomla_4.0.0-beta8-dev+pr.33230-Development-Update_Package.zip, go to System -> Update -> Joomla to update. Then try to use Mail Templates to see if it is working OK. Hopefully it is easier to use than the default implementation.

### CODE REVIEW

**Comment by brianteeman** [comment]
I think you're going to have to change the button names "reset to default" so that it is clearer what they refer to. eg body and message

**Comment by brianteeman** [comment]
And there are now three language strings that can be removed as the fields they refer to are no longer present

**Comment by brianteeman** [comment]
First impressions are great - massive massive improvement

**Comment by joomdonation** [comment]
> First impressions are great - massive massive improvement

Thanks for the feedback. Assume this PR is in right direction, I will work on further clean up and improvements.

**Comment by brianteeman** [comment]
Great - let me know when its ready for full testing

File: administrator/components/com_mails/tmpl/templates/default.php [hunk 1]
```diff
@@ -44,12 +44,11 @@
 								<th scope="col" class="w-15 d-none d-md-table-cell">
 									<?php echo Text::_('COM_MAILS_HEADING_COMPONENT'); ?>
 								</th>
-								<th scope="col" class="w-10 d-none d-md-table-cell">
-									<?php echo Text::_('COM_MAILS_HEADING_TEMPLATES_FOR_LANGUAGES'); ?>
-								</th>
-								<th scope="col" class="w-10 d-none d-md-table-cell">
-									<?php echo Text::_('COM_MAILS_HEADING_NO_TEMPLATES_FOR_LANGUAGES'); ?>
+								<?php if (count($this->languages) > 1) : ?>
+								<th scope="col" class="w-10 d-none d-md-table-cell text-center">
```
Comment (Quy) [inline administrator/components/com_mails/tmpl/templates/default.php]:
```suggestion
								<th scope="col" class="w-10 text-center">
```

File: administrator/components/com_mails/tmpl/templates/default.php [hunk 2]
```diff
@@ -65,49 +64,26 @@
...
-											<?php if ($language->image) : ?>
-												<?php echo HTMLHelper::_('image', 'mod_languages/' . $language->image . '.gif', $language->title_native, array('title' => $language->title_native), true); ?>
-											<?php else : ?>
-												<span class="badge bg-secondary"><?php echo $language->lang_code; ?></span>
-											<?php endif; ?>
-										<?php endif; ?>
-									<?php endforeach; ?>
-								</td>
+								<?php if (count($this->languages) > 1) : ?>
+									<td class="d-none d-md-table-cell text-center">
```
Comment (Quy) [inline administrator/components/com_mails/tmpl/templates/default.php]:
```suggestion
									<td class="text-center">
```

**Comment by joomdonation** [comment]
@Quy Please help committing the suggested changes yourself. Many thanks !

**Comment by joomdonation** [comment]
My last commit added a new column **extension** column to #__mail_templates table. The only improvement from this commit is that it allows hiding mail templates from disabled extensions from Mail Templates management screen. Do you think this is useful/needed? If not, I will revert that commit.

File: administrator/components/com_admin/sql/updates/postgresql/4.0.0-2021-04-22.sql [hunk 3]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE "#__extensions" ADD COLUMN "extension" VARCHAR(127) NOT NULL DEFAULT '' AFTER "template_id";
```
Comment (richard67) [inline administrator/components/com_admin/sql/updates/postgresql/4.0.0-2021-04-22.sql]:
```suggestion
ALTER TABLE "#__extensions" ADD COLUMN "extension" VARCHAR(127) NOT NULL DEFAULT '';
```
PostgreSQL doesn't support "BEFORE" or "AFTER" in "ALTER TABLE ... ADD COLUMN" statements.

File: installation/sql/postgresql/supports.sql [hunk 4]
```diff
@@ -436,29 +437,29 @@ CREATE INDEX "#__mail_templates_idx_language" ON "#__mail_templates" ("language"
...
-('com_users.registration.user.admin_activated', '', 'COM_USERS_EMAIL_ACTIVATED_BY_ADMIN_ACTIVATION_SUBJECT', 'COM_USERS_EMAIL_ACTIVATED_BY_ADMIN_ACTIVATION_BODY', '', '', '{"tags":["name","sitename","siteurl","username"]}'),
-('com_users.registration.admin.verification_request', '', 'COM_USERS_EMAIL_ACTIVATE_WITH_ADMIN_ACTIVATION_SUBJECT', 'COM_USERS_EMAIL_ACTIVATE_WITH_ADMIN_ACTIVATION_BODY', '', '', '{"tags":["name","sitename","email","username","activate"]}'),
-('plg_system_privacyconsent.request.reminder', '', 'PLG_SYSTEM_PRIVACYCONSENT_EMAIL_REMIND_SUBJECT', 'PLG_SYSTEM_PRIVACYCONSENT_EMAIL_REMIND_BODY', '', '', '{"tags":["sitename","url","tokenurl","formurl","token"]}'),
-('com_messages.new_message', '', 'COM_MESSAGES_NEW_MESSAGE', 'COM_MESSAGES_NEW_MESSAGE_BODY', '', '', '{"tags":["subject","message","fromname","sitename","siteurl","fromemail","toname","toemail"]}');
+INSERT INTO "#__mail_templates" ("template_id", "template_id", "language", "subject", "body", "htmlbody", "attachments", "params") VALUES
```
Comment (richard67) [inline installation/sql/postgresql/supports.sql]:
```suggestion
INSERT INTO "#__mail_templates" ("template_id", "extension", "language", "subject", "body", "htmlbody", "attachments", "params") VALUES
```

File: installation/sql/postgresql/supports.sql [hunk 5]
```diff
@@ -421,12 +421,13 @@ CREATE INDEX "#__webauthn_credentials_user_id" ON "#__webauthn_credentials" ("us
...
 CREATE TABLE IF NOT EXISTS "#__mail_templates" (
   "template_id" varchar(127) NOT NULL DEFAULT '',
+  "extension" varchar(127) NOT NULL DEFAULT '',
   "language" char(7) NOT NULL DEFAULT '',
   "subject" varchar(255) NOT NULL DEFAULT '',
   "body" TEXT NOT NULL,
   "htmlbody" TEXT NOT NULL,
   "attachments" TEXT NOT NULL,
-  "params" TEXT NOT NULL,
+  "params" TEXT NOT NULL,
```
Comment (richard67) [inline installation/sql/postgresql/supports.sql]:
```suggestion
  "params" TEXT NOT NULL,
```

File: installation/sql/mysql/supports.sql [hunk 6]
```diff
@@ -412,42 +412,43 @@ CREATE TABLE IF NOT EXISTS `#__webauthn_credentials` (
...
 CREATE TABLE IF NOT EXISTS `#__mail_templates` (
   `template_id` VARCHAR(127) NOT NULL DEFAULT '',
+  `extension` VARCHAR(127) NOT NULL DEFAULT '',
   `language` char(7) NOT NULL DEFAULT '',
   `subject` VARCHAR(255) NOT NULL DEFAULT '',
   `body` TEXT NOT NULL,
   `htmlbody` TEXT NOT NULL,
   `attachments` TEXT NOT NULL,
-  `params` TEXT NOT NULL,
+  `params` TEXT NOT NULL,
```
Comment (richard67) [inline installation/sql/mysql/supports.sql]:
```suggestion
  `params` TEXT NOT NULL,
```

File: administrator/components/com_admin/sql/updates/postgresql/4.0.0-2021-04-22.sql (hunk 3 above)
Comment (richard67) [reply administrator/components/com_admin/sql/updates/postgresql/4.0.0-2021-04-22.sql]: As agreed on Glip I will commit my suggestions. Tuan has seen them.

File: administrator/components/com_admin/sql/updates/mysql/4.0.0-2021-04-22.sql [hunk 7]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE `#__mail_templates` ADD COLUMN `extension` varchar(127) NOT NULL DEFAULT '' AFTER `template_id`;
+UPDATE `#__mail_templates` SET `extension` = SUBSTRING(`template_id`, 1, POSITION('.' IN `template_id`) - 1);
```
Comment (richard67) [inline administrator/components/com_admin/sql/updates/mysql/4.0.0-2021-04-22.sql:2]:
```suggestion
UPDATE `#__mail_templates` SET `extension` = SUBSTRING(`template_id`, 1, POSITION('.' IN `template_id`) - 1);

```

File: administrator/components/com_admin/sql/updates/postgresql/4.0.0-2021-04-22.sql [hunk 8]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE "#__extensions" ADD COLUMN "extension" VARCHAR(127) NOT NULL DEFAULT '';
+UPDATE "#__mail_templates" SET "extension" = SUBSTRING("template_id", 1, POSITION('.' IN "template_id") - 1);
```
Comment (richard67) [inline administrator/components/com_admin/sql/updates/postgresql/4.0.0-2021-04-22.sql:2]:
```suggestion
UPDATE "#__mail_templates" SET "extension" = SUBSTRING("template_id", 1, POSITION('.' IN "template_id") - 1);

```

**Comment by richard67** [comment]
> My last commit added a new column **extension** column to #__mail_templates table. The only improvement from this commit is that it allows hiding mail templates from disabled extensions from Mail Templates management screen. Do you think this is useful/needed?

I think that could be useful, but I am not the expert. @Hackwar What do you think?

**Comment by brianteeman** [comment]
Applied the PR and then the database fix
Opened the component and it said No results
Verified that the database table still exists
Tried to use the search filter and selected my one language (english) and got the following 500 error

**500 Column 'extension' in IN/ALL/ANY subquery is ambiguous Column 'extension' in IN/ALL/ANY subquery is ambiguous**

**Comment by joomdonation** [comment]
@brianteeman Thanks for testing. It was too late for me last night and I did not test my last commit carefully enough. I planned to do final test myself today before pinging you for testing but you were too fast :).

Back to the issue:

1. The no results error, I guess it is because one SQL command is not executed when you use database fix. Please run the SQL commands below to your database then try again, it should solve the issue (of course #__ should be replaced with table prefix used for your site)

```sql
UPDATE `#__mail_templates` SET `extension` = SUBSTRING(`template_id`, 1, POSITION('.' IN `template_id`) - 1);
```

2. The error 500 when select a language is a bug.  I fixed it. Please re-apply the patch and test it again. Thanks.

**Comment by brianteeman** [comment]
It was late for me too or I would have spotted the problem as the query was not structural. ;)

**Comment by brianteeman** [comment]
The Reset buttons don't appear to do anything

**Comment by brianteeman** [comment]
image

I guess break-word is the wrong class

**Comment by brianteeman** [comment]
Other than those small things this is much more usable - thank you very much - you're a superstar

**Comment by joomdonation** [comment]
@brianteeman Thanks for testing and giving feedback

1. For break-word class, it was inherited from original code. I remove the class for now
2. For reset buttons, I can only guess that happens because the change to js file is not applied yet? I'm not sure how you applied the patch, maybe you need to run npm ? I usually test it by download and install update for these kind of PRs.

Maybe you can get this zip file, unzip it media folder on your Joomla installation and try again? It contains the complied js files for this PR which I get from update package

[com_mails.zip](https://github.com/joomla/joomla-cms/files/6363304/com_mails.zip)

File: administrator/components/com_admin/sql/updates/postgresql/4.0.0-2021-04-22.sql [hunk 9]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE "#__extensions" ADD COLUMN "extension" VARCHAR(127) NOT NULL DEFAULT '';
```
Comment (richard67) [inline administrator/components/com_admin/sql/updates/postgresql/4.0.0-2021-04-22.sql]:
```suggestion
ALTER TABLE "#__mail_templates" ADD COLUMN "extension" VARCHAR(127) NOT NULL DEFAULT '';
```

**Comment by brianteeman** [comment]
Sorry I forgot to run npm when i switched branches. The reset buttons work as intended

**Comment by joomdonation** [comment]
> Sorry I forgot to run npm when i switched branches. The reset buttons work as intended

Great. Thanks for confirming.

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on e9cbb4319345e412e4cc2e343f44568f37c42aa1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33230">issues.joomla.org/tracker/joomla-cms/33230</a>.</sub>

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on e9cbb4319345e412e4cc2e343f44568f37c42aa1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33230">issues.joomla.org/tracker/joomla-cms/33230</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33230">issues.joomla.org/tracker/joomla-cms/33230</a>.</sub>

**Comment by joomdonation** [comment]
Thanks everyone  !

**Pull request merged**